### PR TITLE
Retry if RPM lock is temporarily unavailable

### PR DESCRIPTION
**Pull request by witekest**
### What does this PR do?

Wait and retry if Zypper fails acquiring RPM lock file.

### What issues does this PR fix or reference?
Fixes: SUSE/spacewalk#18163

### Previous Behavior
Reportedly highstate was failing in case when RPM lock file was used by some other process.

### New Behavior
Wait and retry if Zypper fails acquiring RPM lock file.

### Merge requirements satisfied?
- [ ] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
No

### CODE REVIEW

File: salt/modules/zypperpkg.py [hunk 1]
```diff
@@ -240,6 +242,12 @@ def _is_lock(self):
         """
         return self.exit_code == self.LOCK_EXIT_CODE
 
+    def _is_rpm_lock(self):
+        """
+        Is this an RPM lock error?
+        """
+        return self.RPM_LOCK_ERR_MESSAGE in self.error_msg
```
Comment (meaksh) [inline salt/modules/zypperpkg.py:249]: We need to be more strict here to consider the error is actually an RPM lock error.

If we only check if that message exists as part of the error coming from Zypper, we can get the minion into and infitine loop when passing that string as a malicious package name, like:

```
# salt-call pkg.install "can't create transaction lock on"
```

**Review by meaksh — CHANGES_REQUESTED** [review]
Thanks for this PR @witekest ! :+1:

I did some comments spotting some things I think we should improve.

File: salt/modules/zypperpkg.py (hunk 1 above)
Comment (vzhestkov) [reply salt/modules/zypperpkg.py:249]: With the other locale on the system the message could be different

Comment (witekest) [reply salt/modules/zypperpkg.py:249]: The problem is the exit code from RPM does not specify what kind of error has actually occurred. Exit code is equal to the number of errors of any kind which have happened during the execution.
What we could do here is to simulate RPM behavior and try to acquire the file lock. Would that be the preferred approach?

**Comment by witekest** [comment]
I've split now the handling of Zypper and RPM locks as the latter does not store information about the locking process. I've tested it by manually acquiring exclusive lock over the RPM lock file. Works as expected and retries the installation until the lock gets released.

File: tests/unit/modules/test_zypperpkg.py [hunk 2]
```diff
@@ -94,7 +94,7 @@ def test_list_upgrades(self):
         }
         with patch.dict(
             zypper.__salt__, {"cmd.run_all": MagicMock(return_value=ref_out)}
-        ):
+        ), patch.object(zypper.__zypper__, "_is_rpm_lock", return_value=False):
```
Comment (meaksh) [inline tests/unit/modules/test_zypperpkg.py:97]: I guess this should be like:
```suggestion
        ), patch.object(zypper.__zypper__, "_is_rpm_lock", MagicMock(return_value=False)):
```

as `return_value` is a parameter for `MagicMock`, right?

File: tests/unit/modules/test_zypperpkg.py [hunk 3]
```diff
@@ -194,7 +194,8 @@ def __call__(self, *args, **kwargs):
             ' type="error">Booya!</message></stream>'
         )
         sniffer = RunSniffer(stdout=stdout_xml_snippet, retcode=1)
-        with patch.dict("salt.modules.zypperpkg.__salt__", {"cmd.run_all": sniffer}):
+        with patch.dict("salt.modules.zypperpkg.__salt__", {"cmd.run_all": sniffer}), \
+            patch.object(zypper.__zypper__, "_is_rpm_lock", return_value=False):
```
Comment (meaksh) [inline tests/unit/modules/test_zypperpkg.py:198]: Same here:
```suggestion
            patch.object(zypper.__zypper__, "_is_rpm_lock", MagicMock(return_value=False)):
```

File: tests/unit/modules/test_zypperpkg.py [hunk 4]
```diff
@@ -241,7 +242,7 @@ def test_list_upgrades_error_handling(self):
         with patch.dict(
             "salt.modules.zypperpkg.__salt__",
             {"cmd.run_all": MagicMock(return_value=ref_out)},
-        ):
+        ), patch.object(zypper.__zypper__, "_is_rpm_lock", return_value=False):
```
Comment (meaksh) [inline tests/unit/modules/test_zypperpkg.py:245]:
```suggestion
        ), patch.object(zypper.__zypper__, "_is_rpm_lock", MagicMock(return_value=False)):
```

File: tests/unit/modules/test_zypperpkg.py [hunk 5]
```diff
@@ -228,7 +229,7 @@ def test_list_upgrades_error_handling(self):
...
```
Comment (meaksh) [inline tests/unit/modules/test_zypperpkg.py:232]: Same here:
```suggestion
        ), patch.object(zypper.__zypper__, "_is_rpm_lock", MagicMock(return_value=False)):
```

File: tests/unit/modules/test_zypperpkg.py (hunk 2 above)
Comment (meaksh) [reply tests/unit/modules/test_zypperpkg.py:97]: Good to know, thanks for clarifying! :+1:

**Comment by meaksh** [comment]
@witekest it seems these changes causes some failures in "windows" as `fcntl` is not available.

We should try to avoid this, for example by doing "try/except" on the import as it is done here: https://github.com/saltstack/salt/blob/master/salt/utils/files.py#L26-L32

**Comment by witekest** [comment]
> @witekest it seems these changes causes some failures in "windows" as `fcntl` is not available.
> 
> We should try to avoid this, for example by doing "try/except" on the import as it is done here: https://github.com/saltstack/salt/blob/master/salt/utils/files.py#L26-L32

Thanks. Good hint. Updated the code accordingly.

**Review by meaksh — DISMISSED** [review]
Looks good! Thanks! :+1:

**Comment by witekest** [comment]
The users have reportedly validated the code by testing it in their environment. Feedback was positive.

**Comment by nicholasmhughes** [comment]
re-run pr-ubuntu-2204-amd64-py3-m2crypto-pytest

**Comment by nicholasmhughes** [comment]
@witekest it looks like pre-commit is failing isort and black checks. Please check the [Contributing documentation](https://docs.saltproject.io/en/latest/topics/development/contributing.html#pre-commit-and-nox-setup) for information on installing and using pre-commit. Thank you!

**Review by twangboy — DISMISSED** [review]

**Review by twangboy — APPROVED** [review]

**Review by Ch3LL — APPROVED** [review]

**Comment by witekest** [comment]
Thanks!

**Pull request merged**